File: src/test/java/org/algohub/engine/judge/MemoryJavaCompilerTest.java
package org.algohub.engine.judge;

import org.junit.Test;

import java.lang.reflect.Method;

import static org.assertj.core.api.Assertions.assertThat;

public class MemoryJavaCompilerTest {

    private static final String SOURCE_CODE =
            "public final class Solution {\n" +
                    "public static String greeting(String name) {\n" +
                    "\treturn \"Hello \" + name;\n" + "}\n}\n";

    @Test
    public void compileAndRunStaticMethod() throws Throwable {

        Object[] greetingObject = MemoryJavaCompiler.INSTANCE.compileMethod(
                "Solution", "greeting", SOURCE_CODE
        );

        Object obj = greetingObject[0];
        Method greeting = (Method) greetingObject[1];

        Object result = greeting.invoke(obj, "Julia");
        assertThat(result).isEqualTo("Hello Julia");
    }

    @Test(expected = NoSuchMethodError.class)
    public void throwsNoSuchMethodExceptionWhenInvokingNonExistingMethod() throws Exception {
        MemoryJavaCompiler.INSTANCE.compileMethod(
                "Solution", "dummy", SOURCE_CODE
        );
    }

    @Test(expected = IllegalStateException.class)
    public void throwsIllegalStateExceptionWhenTryingToInvokeClassWithPrivateConstructor() throws Exception {
        String sourceCode =
                "public final class Solution { private Solution() { }; public void dummy() {} }";

        MemoryJavaCompiler.INSTANCE.compileMethod(
                "Solution", "dummy", sourceCode
        );
    }

    @Test(expected = CompileErrorException.class)
    public void throwsCompileErrorExceptionWhenSourceCodeDoesNotCompile() throws Exception {
        String sourceCodeMissingReturnString =
                "public final class Solution { private Solution() { }; public String dummy() {} }";

        MemoryJavaCompiler.INSTANCE.compileMethod(
                "Solution", "dummy", sourceCodeMissingReturnString
        );
    }
}

File: src/test/kotlin/org/algohub/engine/judge/MemoryJavaCompilerTest.kt
package org.algohub.engine.judge

import org.algohub.engine.compile.CompileErrorException
import org.algohub.engine.compile.MemoryJavaCompiler
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class MemoryJavaCompilerTest {

    @Test
    @Throws(Throwable::class)
    fun compileAndRunStaticMethod() {

        val (instance, method) = MemoryJavaCompiler().compileMethod(
                "Solution", "greeting", SOURCE_CODE
        )

        val result = method.invoke(instance, "Julia")
        assertThat(result).isEqualTo("Hello Julia")
    }

    @Test(expected = NoSuchMethodError::class)
    @Throws(Exception::class)
    fun throwsNoSuchMethodExceptionWhenInvokingNonExistingMethod() {
        MemoryJavaCompiler().compileMethod(
                "Solution", "dummy", SOURCE_CODE
        )
    }

    @Test(expected = IllegalStateException::class)
    @Throws(Exception::class)
    fun throwsIllegalStateExceptionWhenTryingToInvokeClassWithPrivateConstructor() {
        val sourceCode = "public final class Solution { private Solution() { }; public void dummy() {} }"

        MemoryJavaCompiler().compileMethod(
                "Solution", "dummy", sourceCode
        )
    }

    @Test(expected = CompileErrorException::class)
    @Throws(Exception::class)
    fun throwsCompileErrorExceptionWhenSourceCodeDoesNotCompile() {
        val sourceCodeMissingReturnString = "public final class Solution { private Solution() { }; public String dummy() {} }"

        MemoryJavaCompiler().compileMethod(
                "Solution", "dummy", sourceCodeMissingReturnString
        )
    }

    companion object {
        private val SOURCE_CODE =
                "public final class Solution {\npublic static String greeting(String name) {\n\treturn \"Hello \" + name;\n}\n}\n"
    }
}
